opm/simulators/flow/FIBlackoilModel.hpp

```diff
@@ -55,32 +59,74 @@ class FIBlackOilModel : public BlackOilModel<TypeTag>
+    // and order of elements is fixed, and is therefore constrained to only work with CpGrid.
+    // For example, ALUGrid supports refinement and can not assume that.
+    static constexpr bool gridIsUnchanging = std::is_same_v<GetPropType<TypeTag, Properties::Grid>, Dune::CpGrid>;
+
 public:
     FIBlackOilModel(Simulator& simulator)
         : BlackOilModel<TypeTag>(simulator)
     {
+        if constexpr (gridIsUnchanging) {
+            const auto& gv = this->gridView_;
+#ifdef _OPENMP
+            const int nt = omp_get_max_threads();
+            if (nt > 1) {
+                const auto num_elements = gv.size(0);
+                constexpr int max_chunk_size = 1000;
+                const int chunk_size = std::clamp(num_elements / nt, 1, max_chunk_size);
+                OpmLog::debug("Using chunk size " + std::to_string(chunk_size) +
+                              " for property evaluation with " + std::to_string(nt) + " OpenMP threads.");
+                grid_chunk_iterators_.reserve(num_elements / chunk_size + 2);
+                auto it = gv.template begin<0>();
+                const auto end = gv.template end<0>();
+                for (int count = 0; it != end; ++it, ++count) {
+                    if (count % chunk_size == 0) {
+                        grid_chunk_iterators_.push_back(it);
+                    }
+                }
+                grid_chunk_iterators_.push_back(end);
+            } else
+#endif
+            {
+                // With one thread, or without OpenMP, we use a single chunk.
+                grid_chunk_iterators_.push_back(gv.template begin<0>());
+                grid_chunk_iterators_.push_back(gv.template end<0>());
+            }
+        }
     }
 
     void invalidateAndUpdateIntensiveQuantities(unsigned timeIdx) const
     {
-
         this->invalidateIntensiveQuantitiesCache(timeIdx);
-        OPM_BEGIN_PARALLEL_TRY_CATCH()
-        // loop over all elements...
-        ThreadedEntityIterator<GridView, /*codim=*/0> threadedElemIt(this->gridView_);
+        if constexpr (gridIsUnchanging) {
+            OPM_BEGIN_PARALLEL_TRY_CATCH();
+            const int num_chunks = grid_chunk_iterators_.size() - 1;
 #ifdef _OPENMP
-#pragma omp parallel
+#pragma omp parallel for
 #endif
-        {
+            for (int chunk = 0; chunk < num_chunks; ++chunk) {
+                ElementContext elemCtx(this->simulator_);
+                for (auto it = grid_chunk_iterators_[chunk]; it != grid_chunk_iterators_[chunk+1]; ++it) {
+                    const Element& elem = *it;
+                    elemCtx.updatePrimaryStencil(elem);
+                    elemCtx.updatePrimaryIntensiveQuantities(timeIdx);
+                }
+            }
+            OPM_END_PARALLEL_TRY_CATCH("InvalideAndUpdateIntensiveQuantities: state error", this->simulator_.vanguard().grid().comm());
+        } else {
+            // Grid is possibly refined or otherwise changed between calls.
```

Comment:
Do we not need `PARALLEL_TRY_CATCH` in this case?  Do we end up in this branch only in sequential runs?

Reply:
You are right, will move it so that it covers all cases.